src/overmap.cpp

```diff
@@ -4244,95 +4227,63 @@ void overmap::place_specials_pass( std::vector<std::pair<const overmap_special *
+    // Bail out early if we have nothing to place.
+    if( enabled_specials.empty() ) {
+        return;
     }
-    // Make random permutations.
-    std::random_shuffle( mandatory.begin(), mandatory.end() );
-    std::random_shuffle( optional.begin(), optional.end() );
     std::vector<point> sectors = get_sectors();
 
-    // First a regular mandatory pass
-    place_specials_pass( mandatory, sectors, true );
-    // Keep the mandatory vector to inform player about limits properly
-    // Otherwise unplaced vector can be empty while the map is invalid, leading to WTF
-    std::vector<std::pair<const overmap_special *, int>> mandatory_unlimited = mandatory;
-    if( mandatory.empty() ) {
-        current_validity = overmap_valid::valid;
-    } else if( minimum_validity >= overmap_valid::unlimited ) {
-        const std::string unplaced = enumerate_as_string( mandatory.begin(), mandatory.end(),
-        []( const std::pair<const overmap_special *, int> &elem ) {
-            return string_format( "%s (%d)", elem.first->id.c_str(), elem.second );
-        } );
-        dbg( D_WARNING ) << string_format( "couldn't place mandatory overmap specials during first pass (with city checks): %s.", unplaced.c_str() );
-
-        // Couldn't place mandatory specials, try without city checks
-        place_specials_pass( mandatory, sectors, false );
-        current_validity = overmap_valid::unlimited;
-    }
-
-    place_specials_pass( optional, sectors, true );
-
-    unplaced_mandatory_specials.clear();
-    if( !mandatory.empty() ) {
-        current_validity = overmap_valid::invalid;
-    }
-
-    const auto &unplaced = mandatory.empty() ? mandatory_unlimited : mandatory;
-    if( !unplaced.empty() ) {
-        const std::string unplaced_s = enumerate_as_string( mandatory.begin(), mandatory.end(),
-        []( const std::pair<const overmap_special *, int> &elem ) {
-            return string_format( "%s (%d)", elem.first->id.c_str(), elem.second );
-        } );
-        dbg( D_WARNING ) << string_format( "couldn't place mandatory overmap specials: %s.", unplaced_s.c_str() );
-        std::transform( unplaced.begin(), unplaced.end(), std::back_inserter( unplaced_mandatory_specials ),
-        []( const std::pair<const overmap_special *, int> &pr ) {
-            return pr.first;
-        } );
+    // First insure that all minimum instance counts are met.
+    place_specials_pass( enabled_specials, sectors, false );
+    if( std::any_of( enabled_specials.begin(), enabled_specials.end(),
+                     []( overmap_special_placement placement ) {
+                         return placement.instances_placed < placement.special_details->occurrences.min;
+                     } ) ) {
+        // Randomly select from among the nearest uninitialized overmap positions.
+        int previous_distance = 0;
+        std::vector<point> nearest_candidates;
+        for( point candidate_addr : closest_points_first( 10, pos() ) ) {
```

Comment:
I don't have evidence that it breaks, but it will be hard to gather it if it avoids displaying the errors until the last moment (ie. searching for mission).
We will have to ask people to make sure their overmaps aren't migrated from old worldgen, then there will be a painful process of reproducing the error with no real tools for it.
Then, when new specials are added, there will be no surefire method of checking if they don't break, only the probabilistic unit test.

Then there are mods and non-standard settings, which aren't covered under unit tests at all, meaning they are at huge risk of breaking with no good info.